File: app/database/rss_scraper/rss_scraper.js
const Parser = require('rss-parser');
const readline = require('readline');
const { Pool, Client } = require('pg');

const client = new Client({
    user: 'brojogan',
    host: 'polybius.bowdens.me',
    database: 'ultracast',
    password: 'GbB8j6Op',
    port: 5432
});

let count=0;
const parser = new Parser();
let rl;

client.connect(err => {
    if (err) {
        console.log("connection error");
    } else {
        console.log("database ready");
        rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        rl.on('close', () => {
            console.log("done reading; there will be more async resquests tho");

        });
        rl.on('line', (url) => {
            async function foo() {
                let podcastid = -1;
                console.log(" -- reading url '" + url + "'");
                let feed;
                try {
                    feed = await parser.parseURL(url);
                } catch (error) {
                    console.error("there was an error reading the rss feed " + url + ". continuing");
                    console.error(error);
                    return;
                }
                console.log("   -- got feed for url '" + url + "'");
                try {
                    let res = await client.query("insert into podcasts (rssfeed, title, author, description, thumbnail) values ($1, $2, $3, $4, $5) returning id",
                                 [url, feed.title, feed.itunes ? feed.itunes.author : null, feed.description, feed.itunes ? feed.itunes.image : null]
                                );
                    if (res.rowCount != 1) {
                        console.error("    -- something went wrong inserting into podcasts:");
                        console.error(res);
                    } else {
                        //console.log("   -- added url '" + url + "'");
                        podcastid = res.rows[0].id;
                    }
                } catch (error) {
                    console.error("    -- there was an error inserting");
                    console.error(error);
                    return;
                }
                let categories = [];
                if (feed.itunes) {
                    categories = feed.itunes.categories;
                    //console.log("    -- categories are:");
                    //console.log(categories);
                    try {
                        for (let a of categories) {}
                    } catch (error) {
                        console.error("error iterating: ");
                        console.error(error);
                    }
                }
                console.log(" for " + url + " categories is " + categories);
                for (let category of categories) {
                    try {
                        let res = await client.query("select * from categories where name = $1",
                                                     [category]
                                                    );
                        //console.log(category);
                        let categoryid;
                        if (res.rowCount == 0) {
                            try {
                                let newres = await client.query("insert into categories (name) values ($1) returning id",
                                                            [category]
                                                           );
                                //console.log("after inserting new category newres is");
                                //console.log(newres);
                                //console.log("newres.rows: " + newres.rows);
                                //console.log("newres.rows[0]: " + newres.rows[0]);
                                //console.log("newres.rows[0].id: " + newres.rows[0].id);

                                categoryid = newres.rows[0].id;
                            } catch (error) {
                                console.error("    -- something went wrong inserting the new category");
                                console.error(error);
                            }
                        } else {
                            //console.log("after selecting category, res is");
                            //console.log(res);
                            //console.log("res.rows: " + res.rows);
                            //console.log("res.rows[0]: " + res.rows[0]);
                            //console.log("res.rows[0].id: " + res.rows[0].id);
                            categoryid = res.rows[0].id;
                        }
                        try {
                            let newres = await client.query("insert into podcastCategories (podcastId, categoryId) values ($1, $2)",
                                                           [podcastid, categoryid]);
                        } catch (error) {
                            console.error("    -- something went wrong inserting the new podcast category");
                            console.error("      -- podcastid was " + podcastid + ", categoryid was " + categoryid);
                            console.error(error);
                            return;
                        }
                    } catch (error) {
                        console.error("    -- there was an error checking the categories");
                        return;
                    }
                }
            }
            count += 1;
            console.log("count: " + count);
            foo();
        });
    }
});

File: app/database/rss_scraper/rss_scraper.py
import feedparser
import sys
import psycopg2
import socket

def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)

test=False
socket.setdefaulttimeout(10)

try:
    conn = psycopg2.connect(dbname="ultracast", user="brojogan", password="GbB8j6Op", host="polybius.bowdens.me", port=5432)
except Exception as e:
    eprint(e)
    exit(1)

cur = conn.cursor()

count = 0

for line in sys.stdin:
    url = line.rstrip()
    eprint(url)
    data = {}
    try:
        data = feedparser.parse(url)
    except Exception as e:
        eprint(" - error parsing:")
        eprint(e)
        print(url)
        continue
    if data["bozo"]:
        eprint(" - malformed rss, continuing")
        print(url)
        continue
    try:
        cur.execute("""
            insert into podcasts
            (rssfeed, title, author, description, thumbnail)
            values (%s, %s, %s, %s, %s)
            returning id;
        """,
        (
            url,
            data.feed.get("title"),
            data.feed.get("author"),
            data.feed.get("description"),
            data.feed.get("image")["href"] if data.feed.get("image") else None
        ))
    except Exception as e:
        conn.rollback()
        eprint(" - error inserting into podcasts:")
        eprint(e)
        print(url)
        continue
    podcastid = cur.fetchone()[0]
    eprint("  inserted podcast with id {}".format(podcastid))

    categories = [data.feed.get("category")] if data.feed.get("category") else []
    #categories = [x.term for x in data.feed.get("tags",[]) if x.scheme == 'http://www.itunes.com/dtds/podcast-1.0.dtd']
    eprint(" - categories are: {}".format(categories))

    for category in categories:
        eprint("    - category: {}".format(category))
        try:
            cur.execute("""
                select id from categories
                where name = %s;
            """,
            (category,)
            )
        except Exception as e:
            eprint(" -- error getting category {}".format(category))
            eprint(e)
            continue
        categoryid = None
        res = cur.fetchone()
        if res is None:
            # we need to insert a new category
            try:
                cur.execute("""
                    insert into categories (name)
                    values (%s)
                    returning id
                    """,
                    (category,)
                )
            except Exception as e:
                conn.rollback()
                eprint(" -- error inserting new category {}".format(category))
                eprint(e)
                continue
            categoryid = cur.fetchone()[0]
            eprint("  inserted category '{}' with id {}".format(category ,categoryid))
        else:
            categoryid = res[0]
            eprint("  got category '{}' with id {}".format(category ,categoryid))

        # insert podcastCategories
        try:
            cur.execute("""
                insert into podcastCategories (podcastId, categoryId)
                values (%s, %s)
                """,
                (podcastid, categoryid)
            )
            eprint("  inserted podcast category with ({},{})".format(podcastid, categoryid))
        except Exception as e:
            eprint(" -- error inserting podcast category ({}, {})".format(podcastid, categoryid))
            eprint(e)
            continue
    eprint("finished {}".format(url))
    if test:
        conn.rollback()
        eprint("rolled back because in test mode")
    else:
        conn.commit()
        eprint("committed")
    count += 1
    eprint("line {} finished\n".format(count))


cur.close()
conn.close()
